src/main/java/com/laserfiche/repository/api/RepositoryApiClientImpl.java

```diff
@@ -48,13 +41,34 @@ protected RepositoryApiClientImpl(String servicePrincipalKey, AccessKey accessKe
         templateDefinitionsClient = new TemplateDefinitionsClientImpl(baseUrl, httpClient);
     }
 
-    public static RepositoryApiClient CreateFromAccessKey(String servicePrincipalKey, AccessKey accessKey,
-            String baseUrlDebug) {
-        return new RepositoryApiClientImpl(servicePrincipalKey, accessKey, baseUrlDebug);
+    public static RepositoryApiClient CreateFromHttpRequestHandler(String servicePrincipalKey, AccessKey accessKey,
+            String baseUrl) {
+        String baseUrlDebug = baseUrl != null ? baseUrl : "https://api." + accessKey.getDomain() + "/repository";
+        httpClient
+                .config()
+                .setObjectMapper(new RepositoryClientObjectMapper())
+                .interceptor(new OAuthInterceptor(servicePrincipalKey, accessKey));
+        return new RepositoryApiClientImpl(httpClient, baseUrlDebug);
+    }
+
+    public static RepositoryApiClient CreateFromHttpRequestHandler(String repositoryId, String username,
+            String password,
+            TokenClient client, String baseUrl) {
+        httpClient
+                .config()
+                .setObjectMapper(new RepositoryClientObjectMapper())
+                .interceptor(new SelfHostedInterceptor(repositoryId, username, password, baseUrl, client));
+        return new RepositoryApiClientImpl(httpClient, baseUrl);
     }
 
     public static RepositoryApiClient CreateFromAccessKey(String servicePrincipalKey, AccessKey accessKey) {
-        return CreateFromAccessKey(servicePrincipalKey, accessKey, null);
+        return CreateFromHttpRequestHandler(servicePrincipalKey, accessKey, null);
+    }
+
+    public static RepositoryApiClient CreateFromUsernamePassword(String repoId, String username, String password,
+            String baseUrl) {
+        String baseUrlWithSlash = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.lastIndexOf("/")) : baseUrl;
```

Comment:
Why use `lastIndexOf` if we already know it's at the end. BTW, `lastIndexOf` is O(n) where n is the number of characters in the string.

Reply:
yea so this was put to account for the edge case of having a backslash at the end of the self hosted base URL so for example if i do not have this:
`https://v-dev-w10x64-73.laserfiche.com/LFRepositoryAPI/` will throw an error even tho it's not supposed to